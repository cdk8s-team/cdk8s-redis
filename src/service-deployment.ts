import { Construct } from 'constructs';
import * as k8s from './imports/k8s';

export interface ServiceDeploymentOptions {
  /**
   * @default 1
   */
  readonly replicas?: number;

  /**
   * Container image.
   */
  readonly image: string;

  /**
   * Additional labels to apply to resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };

  /**
   * Pod resource requirements
   * @default - cpu: '100m', memory: '100Mi'
   */
  readonly resources?: k8s.ResourceRequirements;

  /**
   * @default 80
   */
  readonly externalPort?: number;

  /**
   * @default 8080
   */
  readonly containerPort?: number;

  /**
   * Environment variables to pass to the pod
   */
  readonly env?: { [name: string]: string };

  /**
   * The name of the container
   * @default 'app'
   */
  readonly containerName?: string;

  /**
   * The type of service resource.
   * @default ServiceType.CLUSTER_IP
   */
  readonly serviceType?: ServiceType;
}

export enum ServiceType {
  EXTERNAL_NAME = 'ExternalName',
  CLUSTER_IP = 'ClusterIP',
  NODE_PORT = 'NodePort',
  LOAD_BALANCER = 'LoadBalancer'
}

export class ServiceDeployment extends Construct {

  public readonly host: string;

  constructor(scope: Construct, id: string, options: ServiceDeploymentOptions) {
    super(scope, id);

    const replicas = options.replicas ?? 1;
    const resources = options.resources ?? { requests: { cpu: '100m', memory: '100Mi' } };
    const externalPort = options.externalPort ?? 80;
    const containerPort = options.containerPort ?? 8080;
    const containerName = options.containerName ?? 'app';
    const serviceType = options.serviceType ?? ServiceType.CLUSTER_IP;

    const label = {
      ...options.labels,
      app: this.node.addr,
    };

    const service = new k8s.Service(this, 'service', {
      metadata: { labels: label },
      spec: {
        type: serviceType,
        ports: [{ port: externalPort, targetPort: containerPort }],
        selector: label,
      },
    });


    new k8s.Deployment(this, 'deployment', {
      spec: {
        selector: { matchLabels: label },
        replicas: replicas,
        template: {
          metadata: { labels: label },
          spec: {
            containers: [
              {
                name: containerName,
                image: options.image,
                ports: [{ containerPort }],
                resources: resources,
                env: renderEnv(options.env),
              },
            ],
          },
        },
      },
    });

    this.host = service.name;
  }
}

function renderEnv(env: { [key: string]: string } = { }): k8s.EnvVar[] {
  const result = new Array<k8s.EnvVar>();
  for (const [key, value] of Object.entries(env)) {
    result.push({
      name: key,
      value: value,
    });
  }
  return result;
}