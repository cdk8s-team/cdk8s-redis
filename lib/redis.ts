import { Construct } from 'constructs';
import { ServiceDeployment } from './service-deployment';

export interface RedisOptions {
  /**
   * Number of slave replicas.
   * @default 2
   */
  readonly slaveReplicas?: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };
}

export class Redis extends Construct {
  /**
   * The DNS host for the master server.
   */
  public readonly masterHost: string;

  /**
   * The DNS host for the slave service.
   */
  public readonly slaveHost: string;

  constructor(scope: Construct, id: string, options: RedisOptions = { }) {
    super(scope, id);

    const master = new ServiceDeployment(this, 'master', {
      image: 'k8s.gcr.io/redis:e2e',
      containerPort: 6379,
      externalPort: 6379,
      containerName: 'master',
      env: { GET_HOSTS_FROM: 'dns' },
      labels: {
        app: 'redis',
        role: 'master',
        ...options.labels,
      },
    });

    this.masterHost = master.host;

    const slaveReplicas = options.slaveReplicas ?? 2;
    if (slaveReplicas > 0) {
      const slave = new ServiceDeployment(this, 'slave', {
        image: 'gcr.io/google_samples/gb-redisslave:v1',
        containerName: 'slave',
        containerPort: 6379,
        externalPort: 6379,
        env: { GET_HOSTS_FROM: 'env', REDIS_MASTER_SERVICE_HOST: this.masterHost },
        replicas: slaveReplicas,
        labels: {
          app: 'redis',
          role: 'slave',
          ...options.labels,
        },
      });

      this.slaveHost = slave.host;
    } else {
      // if we have no slave, then use the same host as the master
      this.slaveHost = master.host;
    }
  }
}
