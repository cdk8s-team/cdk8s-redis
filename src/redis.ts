import { Construct } from 'constructs';
import { ServiceDeployment } from './service-deployment';

export interface RedisOptions {
  /**
   * Number of replicas.
   * @default 2
   */
  readonly replicas?: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };
}

export class Redis extends Construct {
  /**
   * The DNS host for the primary service.
   */
  public readonly primaryHost: string;

  /**
   * The DNS host for the replica service.
   */
  public readonly replicaHost: string;

  constructor(scope: Construct, id: string, options: RedisOptions = { }) {
    super(scope, id);

    const primary = new ServiceDeployment(this, 'primary', {
      image: 'k8s.gcr.io/redis:e2e',
      containerPort: 6379,
      externalPort: 6379,
      containerName: 'primary',
      env: { GET_HOSTS_FROM: 'dns' },
      labels: {
        app: 'redis',
        role: 'primary',
        ...options.labels,
      },
    });

    this.primaryHost = primary.host;

    const replicas = options.replicas ?? 2;
    if (replicas > 0) {
      const replica = new ServiceDeployment(this, 'replica', {
        image: 'gcr.io/google_samples/gb-redisslave:v1',
        containerName: 'replica',
        containerPort: 6379,
        externalPort: 6379,
        env: { GET_HOSTS_FROM: 'env', REDIS_MASTER_SERVICE_HOST: this.primaryHost },
        replicas: replicas,
        labels: {
          app: 'redis',
          role: 'replica',
          ...options.labels,
        },
      });

      this.replicaHost = replica.host;
    } else {
      // if we have no slave, then use the same host as the primary
      this.replicaHost = primary.host;
    }
  }
}
