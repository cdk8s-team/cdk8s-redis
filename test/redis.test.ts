import { Chart, Testing } from 'cdk8s';
import { Redis } from '../src/redis';

test('defaults', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'test');

  // WHEN
  const redis = new Redis(chart, 'redis');

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
  expect(redis.masterHost).toEqual('test-redis-master-service-c852b917');
  expect(redis.slaveHost).toEqual('test-redis-slave-service-c8fdf1aa');
});

test('no slave replicas will deploy only the master', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'test');

  // WHEN
  const redis = new Redis(chart, 'redis', {
    slaveReplicas: 0,
  });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
  expect(redis.masterHost).toEqual('test-redis-master-service-c852b917');
  expect(redis.masterHost).toEqual(redis.slaveHost); // slave = master
});

test('labels can be added', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'test');

  // WHEN
  new Redis(chart, 'redis', {
    labels: {
      mylabel: '1234',
    },
  });

  // THEN
  const resources = Testing.synth(chart);

  for (const svc of resources.filter(r => r.kind === 'Service')) {
    expect(svc.metadata.labels.mylabel).toEqual('1234');
  }

  for (const dep of resources.filter(r => r.kind === 'Deployment')) {
    expect(dep.spec.selector.matchLabels.mylabel).toEqual('1234');
    expect(dep.spec.template.metadata.labels.mylabel).toEqual('1234');
  }
});
