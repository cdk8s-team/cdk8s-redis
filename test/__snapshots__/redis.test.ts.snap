// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app": "c8b2fe46a20ee748f914a8b108427882c2c06096a9",
        "role": "primary",
      },
      "name": "test-redis-primary-service-c8f36e4b",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 6379,
          "targetPort": 6379,
        },
      ],
      "selector": Object {
        "app": "c8b2fe46a20ee748f914a8b108427882c2c06096a9",
        "role": "primary",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-redis-primary-deployment-c8c0d573",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "c8b2fe46a20ee748f914a8b108427882c2c06096a9",
          "role": "primary",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "c8b2fe46a20ee748f914a8b108427882c2c06096a9",
            "role": "primary",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "GET_HOSTS_FROM",
                  "value": "dns",
                },
              ],
              "image": "k8s.gcr.io/redis:e2e",
              "name": "primary",
              "ports": Array [
                Object {
                  "containerPort": 6379,
                },
              ],
              "resources": Object {
                "requests": Object {
                  "cpu": "100m",
                  "memory": "100Mi",
                },
              },
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app": "c8dfe5fbb37f49bc8b747fb7ebdee0549ee0d2cbc0",
        "role": "replica",
      },
      "name": "test-redis-replica-service-c814a5a4",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 6379,
          "targetPort": 6379,
        },
      ],
      "selector": Object {
        "app": "c8dfe5fbb37f49bc8b747fb7ebdee0549ee0d2cbc0",
        "role": "replica",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-redis-replica-deployment-c8e1d6e2",
    },
    "spec": Object {
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "app": "c8dfe5fbb37f49bc8b747fb7ebdee0549ee0d2cbc0",
          "role": "replica",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "c8dfe5fbb37f49bc8b747fb7ebdee0549ee0d2cbc0",
            "role": "replica",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "GET_HOSTS_FROM",
                  "value": "env",
                },
                Object {
                  "name": "REDIS_MASTER_SERVICE_HOST",
                  "value": "test-redis-primary-service-c8f36e4b",
                },
              ],
              "image": "gcr.io/google_samples/gb-redisslave:v1",
              "name": "replica",
              "ports": Array [
                Object {
                  "containerPort": 6379,
                },
              ],
              "resources": Object {
                "requests": Object {
                  "cpu": "100m",
                  "memory": "100Mi",
                },
              },
            },
          ],
        },
      },
    },
  },
]
`;

exports[`no replicas will deploy only the primary 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app": "c8b2fe46a20ee748f914a8b108427882c2c06096a9",
        "role": "primary",
      },
      "name": "test-redis-primary-service-c8f36e4b",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 6379,
          "targetPort": 6379,
        },
      ],
      "selector": Object {
        "app": "c8b2fe46a20ee748f914a8b108427882c2c06096a9",
        "role": "primary",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-redis-primary-deployment-c8c0d573",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "c8b2fe46a20ee748f914a8b108427882c2c06096a9",
          "role": "primary",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "c8b2fe46a20ee748f914a8b108427882c2c06096a9",
            "role": "primary",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "GET_HOSTS_FROM",
                  "value": "dns",
                },
              ],
              "image": "k8s.gcr.io/redis:e2e",
              "name": "primary",
              "ports": Array [
                Object {
                  "containerPort": 6379,
                },
              ],
              "resources": Object {
                "requests": Object {
                  "cpu": "100m",
                  "memory": "100Mi",
                },
              },
            },
          ],
        },
      },
    },
  },
]
`;
